include/ocvsmd/sdk/execution.hpp

```diff
@@ -0,0 +1,154 @@
+//
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+//
+
+#ifndef OCVSMD_SDK_EXECUTION_HPP_INCLUDED
+#define OCVSMD_SDK_EXECUTION_HPP_INCLUDED
+
+#include "ocvsmd/platform/defines.hpp"
+
+#include <cetl/pf17/cetlpf.hpp>
+
+#include <functional>
+#include <memory>
+
+namespace ocvsmd
+{
+namespace sdk
+{
+
+/// Internal implementation details.
+/// Not supposed to be used directly by the users of the SDK.
+///
+namespace detail
+{
+
+template <typename Result>
+class StateOf;
+
+template <typename Result>
+class ReceiverOf final
+{
+public:
+    explicit ReceiverOf(std::shared_ptr<StateOf<Result>> state)
+        : state_{std::move(state)}
+    {
+    }
+
+    template <typename... Args>
+    void operator()(Args&&... args)
+    {
+        state_->maybe_result_.emplace(std::forward<Args>(args)...);
+    }
+
+private:
+    std::shared_ptr<StateOf<Result>> state_;
+
+};  // ReceiverOf
+
+template <typename Result>
+class StateOf final : public std::enable_shared_from_this<StateOf<Result>>
+{
+public:
+    bool completed() const
+    {
+        return maybe_result_.has_value();
+    }
+
+    Result get()
+    {
+        CETL_DEBUG_ASSERT(maybe_result_, "");
+        return std::move(*maybe_result_);
+    }
+
+    ReceiverOf<Result> makeReceiver()
+    {
+        return ReceiverOf<Result>{this->shared_from_this()};
+    }
+
+private:
+    friend class ReceiverOf<Result>;
+
+    cetl::optional<Result> maybe_result_;
+
+};  // StateOf
+
+}  // namespace detail
+
+/// Abstract interface of a result sender.
```

Comment:
These sender and receivers are inspired by the [P2300](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2024/p2300r10.html) c++ proposal. Not even close of course, but the main idea is that client of our SDK could compose senders (using `.then( callback )` or `|` operator) to build necessary async pipeline, and then `sync_wait` to wait for result (see `src/cli/main.cpp:94`).